### PR TITLE
Python: Add `MarkupSafe` model

### PR DESCRIPTION
**Pull request by RasmusWL**
The escaping concept is inspired by what we have in [Go](https://github.com/github/codeql-go/blob/2a80a60468b11742db146f2cf933ae6d1ca2feaa/ql/src/semmle/go/Concepts.qll#L127-L182), but following the same naming as the encoding/decoding concepts we already have.

Here are some of the interesting decisions around this PR, that are not highlighted as part of the commits/code.

### default taint steps for escaping?

I considered adding escaping for _other_ kinds as additional taint steps, so for example in our SQL injection query, if you do escaping for HTML, that would still be a valid taint step. I opted _not_ to do this, since knowing whether some kind of escaping will make user-input safe for SQL injection is sort of a tricky question... what if you applied a chain of escapers, such as `js_escape(html_escape(ldap_escape(user_input)))`, will it _then_ be safe for SQL? I guess the worst that could happen is that we end up with false-positives, with this approach. So if _you_ think this is a really good idea that we should totally do, let me know!

### isSanitizingStep

Also see the note about not having `isSanitizingStep`. I've talked with @aschackmull about solving this, but we haven't quite agreed on how to actually do it yet. So for now, we can get this PR merged with the less-than-ideal behavior, and fix it down the line :+1:

### CODE REVIEW

File: python/ql/src/semmle/python/frameworks/MarkupSafe.qll [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+   * Provides models for the `markupsafe.Markup` class
+   *
+   * See https://markupsafe.palletsprojects.com/en/2.0.x/escaping/#markupsafe.Markup.
+   */
+  module Markup {
+    /** Gets a reference to the `markupsafe.Markup` class. */
+    API::Node classRef() {
+      result = API::moduleImport("markupsafe").getMember("Markup")
+      or
+      result = API::moduleImport("flask").getMember("Markup")
```
Comment (jorgectf) [inline python/ql/src/semmle/python/frameworks/MarkupSafe.qll:27]:
```suggestion
      result = API::moduleImport(["markupsafe", "flask"]).getMember("Markup")
```

File: python/ql/src/semmle/python/frameworks/MarkupSafe.qll [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    class StringConcat extends Markup::InstanceSource, DataFlow::CfgNode {
+      override BinaryExprNode node;
+
+      StringConcat() {
+        node.getOp() instanceof Add and
+        (
+          instance().asCfgNode() = node.getLeft()
+          or
+          instance().asCfgNode() = node.getRight()
+        )
```
Comment (jorgectf) [inline python/ql/src/semmle/python/frameworks/MarkupSafe.qll:69]:
```suggestion
        instance().asCfgNode() in [node.getLeft(), node.getRight()]
```

File: python/ql/src/semmle/python/frameworks/MarkupSafe.qll (hunk 1 above)
Comment (RasmusWL) [reply python/ql/src/semmle/python/frameworks/MarkupSafe.qll:27]: Although you certainly could, I think having these one separate lines makes the code easier to quickly scan with your eyes.

File: python/ql/src/semmle/python/frameworks/MarkupSafe.qll (hunk 2 above)
Comment (RasmusWL) [reply python/ql/src/semmle/python/frameworks/MarkupSafe.qll]: Thanks :tada:

File: python/ql/src/semmle/python/frameworks/MarkupSafe.qll (hunk 1 above)
Comment (jorgectf) [reply python/ql/src/semmle/python/frameworks/MarkupSafe.qll:27]: You are right :), thought that too with [this block](https://github.com/github/codeql/blob/c77884b8c4ef0525f8acaf8c24d53cfae06e84d5/python/ql/src/semmle/python/frameworks/MarkupSafe.qll#L102-L104).

File: python/ql/src/semmle/python/Concepts.qll [hunk 3]
```diff
@@ -293,6 +293,78 @@ module SqlExecution {
...
+ * extend `Escaping::Range` instead.
+ */
+class Escaping extends DataFlow::Node {
+  Escaping::Range range;
+
+  Escaping() {
+    this = range and
+    // escapes that don't have _both_ input/output defined are not valid
+    exists(range.getAnInput()) and
+    exists(range.getOutput())
```
Comment (tausbn) [inline python/ql/src/semmle/python/Concepts.qll:310]: I'm not quite sure what the function of these two lines is. What would go wrong if there was a "half-defined" instance of `Escaping`?

File: python/ql/src/semmle/python/frameworks/MarkupSafe.qll [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        node.getOp() instanceof Add and
+        instance().asCfgNode() in [node.getLeft(), node.getRight()]
+      }
+    }
+
+    /** A string format with `markupsafe.Markup` as the format string. */
+    class StringFormat extends Markup::InstanceSource, DataFlow::CallCfgNode {
+      StringFormat() {
+        exists(DataFlow::AttrRead attr | this.getFunction() = attr |
+          attr.getAttributeName() = "format" and
```
Comment (tausbn) [inline python/ql/src/semmle/python/frameworks/MarkupSafe.qll:73]: Do we plan on supporting `%`-style formatting?

Also, would this be a place to use `MethodCallNode?

**Review by tausbn — APPROVED** [review]
A couple of questions, but apart from that this looks good to me. :+1:

File: python/ql/src/semmle/python/Concepts.qll (hunk 3 above)
Comment (RasmusWL) [reply python/ql/src/semmle/python/Concepts.qll:310]: The problematic part is that we do `override predicate isSanitizer(DataFlow::Node node) { node = any(HtmlEscaping esc).getOutput() }` in the XSS configuration, and then the example from the commit message of https://github.com/github/codeql/pull/6092/commits/498703fc81727d733265dc0c9002205cfd8b5534

File: python/ql/src/semmle/python/frameworks/MarkupSafe.qll (hunk 4 above)
Comment (RasmusWL) [reply python/ql/src/semmle/python/frameworks/MarkupSafe.qll:73]: `MethodCallNode` for sure :+1: I didn't initially plan on support `%`-style formatting, but we can do so :)

Comment (RasmusWL) [reply python/ql/src/semmle/python/frameworks/MarkupSafe.qll]: https://github.com/github/codeql/pull/6092/commits/c2708176b140e56cfbcc752db03ca59ef45f24e8 adds %-style formatting :+1:

**Review by tausbn — APPROVED** [review]
Cool stuff!

:shipit:

**Pull request merged**